File: src/android/app/src/main/java/org/citra/citra_emu/utils/FileBrowserHelper.java
package org.citra.citra_emu.utils;

import android.content.ClipData;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;

import java.util.ArrayList;
import java.util.List;

public final class FileBrowserHelper {

    @Nullable
    public static String[] getSelectedFiles(Intent result, Context context, List<String> extension) {
        ClipData clipData = result.getClipData();
        List<DocumentFile> files = new ArrayList<>();
        if (clipData == null) {
            files.add(DocumentFile.fromSingleUri(context, result.getData()));
        } else {
            for (int i = 0; i < clipData.getItemCount(); i++) {
                ClipData.Item item = clipData.getItemAt(i);
                Uri uri = item.getUri();
                files.add(DocumentFile.fromSingleUri(context, uri));
            }
        }
        if (!files.isEmpty()) {
            List<String> filePaths = new ArrayList<>();
            for (int i = 0; i < files.size(); i++) {
                DocumentFile file = files.get(i);
                String filename = file.getName();
                int extensionStart = filename.lastIndexOf('.');
                if (extensionStart > 0) {
                    String fileExtension = filename.substring(extensionStart + 1);
                    if (extension.contains(fileExtension)) {
                        filePaths.add(file.getUri().toString());
                    }
                }
            }
            if (filePaths.isEmpty()) {
                return null;
            }
            return filePaths.toArray(new String[0]);
        }

        return null;
    }
}

File: src/android/app/src/main/java/org/citra/citra_emu/utils/FileBrowserHelper.kt
// Copyright 2023 Citra Emulator Project
// Licensed under GPLv2 or any later version
// Refer to the license.txt file included.

package org.citra.citra_emu.utils

import android.content.Context
import android.content.Intent
import androidx.documentfile.provider.DocumentFile

object FileBrowserHelper {
    fun getSelectedFiles(
        result: Intent,
        context: Context,
        extension: List<String?>
    ): Array<String>? {
        val clipData = result.clipData
        val files: MutableList<DocumentFile?> = ArrayList()
        if (clipData == null) {
            files.add(DocumentFile.fromSingleUri(context, result.data!!))
        } else {
            for (i in 0 until clipData.itemCount) {
                val item = clipData.getItemAt(i)
                files.add(DocumentFile.fromSingleUri(context, item.uri))
            }
        }
        if (files.isNotEmpty()) {
            val filePaths: MutableList<String> = ArrayList()
            for (i in files.indices) {
                val file = files[i]
                val filename = file?.name
                val extensionStart = filename?.lastIndexOf('.') ?: 0
                if (extensionStart > 0) {
                    val fileExtension = filename?.substring(extensionStart + 1)
                    if (extension.contains(fileExtension)) {
                        filePaths.add(file?.uri.toString())
                    }
                }
            }
            return if (filePaths.isEmpty()) null else filePaths.toTypedArray<String>()
        }
        return null
    }
}
